Add email verification notice to publish screen

coreh opened the pull request:
After some discussion, we concluded that the ideal solution to #2650 involves disabling the _Publish_ button altogether and displaying a notice to the users right away, rather than delaying it until after they try to publish. (This should provide the least amount of surprise/frustration)

I started off with @gyoshev's PR (#4806, mad props for writing the initial fix for this! 👏). I then modified it by adding the notice and disabling the primary button if not verified. The Dialog with more information/options gets displayed once you click on the notice. Here's how it looks:

![screen shot 2016-05-19 at 18 45 51](https://cloud.githubusercontent.com/assets/418473/15410759/fab545a4-1df1-11e6-9b75-acb465c18be1.png)

![screen shot 2016-05-19 at 18 45 57](https://cloud.githubusercontent.com/assets/418473/15410767/ff939080-1df1-11e6-8844-aa0f02329083.png)

The verification notice is reactive, so once the user has verified their email address it will go away. To do this I used a combination of polling the user API every 15 seconds, and signaling via the `localStorage` API from the confirmation success page.

@shaunandrews can you take a look on the design aspect of this? I also need someone to tweak the copy. I added a paragraph explaining why verification is needed, not sure we want this.

folletto commented:
Looks good!

Tested in Chrome with both "Publish" and "Update" actions. 👍 

Small things:

1. Can we offset the "arrow" pointing to the button to be centered? The reason is that with languages that have longer strings the two buttons will stack one above the other, thus it would just look on one side, and even just in english it's not easy to center in the blue button itself — hence I'd just center it ([mock screenshot](https://cloud.githubusercontent.com/assets/4389/15425842/da6e0cf4-1e82-11e6-945c-8cd164c8d58b.png))
2. When clicked flashes for an instant the focus border, which on an edge-to-edge element looks a bit odd (but I'll wait other opinions on this). If we want to avoid that, a simple way could be to disable the highlight border on `.editor-ground-control__email-verification-notice:focus:hover` (and its pseudo element).
3. Copy: given the description text is a bit long, I'd suggest to move to bullet points the two inline actions, like [this](https://cloud.githubusercontent.com/assets/4389/15426005/b334449a-1e83-11e6-9394-2191039e3b3e.png) (check the bullets padding).
4. Copy: I think we don't use the full stop at the end of titles, so I'd take it away.
5. Copy: I'd change the button label from "Got it" to "Close" or "Ok", since we don't know why the user opened the informations (this is a super super minor detail).

None of these are a blocker imho, but they are fairly straightforward. Check if they make sense. :)

coreh commented:
@shaunandrews made some tweaks to the dialog, so it's not that long.

I also added `text-overflow: ellipsis` to the email, and moved the resend action to a button.

<img width="587" alt="screen shot 2016-05-20 at 5 05 36 pm" src="https://cloud.githubusercontent.com/assets/418473/15440708/5774698a-1ead-11e6-8faf-338f898da234.png">

Tip is now centered:

<img width="283" alt="screen shot 2016-05-20 at 5 09 21 pm" src="https://cloud.githubusercontent.com/assets/418473/15440762/a3b63abc-1ead-11e6-9d76-3ce61089aa5b.png">

(shadow in the screenshot is my dock)

coreh commented:
Now with OK button, too (Had missed that):

<img width="482" alt="screen shot 2016-05-20 at 5 13 24 pm" src="https://cloud.githubusercontent.com/assets/418473/15440870/3924da90-1eae-11e6-889f-01b2f7626589.png">

folletto commented:
Looks great 👍

rickybanister commented:
Is this ready to merge or does it need any more review?

coreh commented:
Added a spinner when resending the email. After reviewing that, should be ready to merge.

shaunandrews commented:
Can we apply this same notice when editing pages?

![screen shot 2016-05-24 at 2 51 35 pm](https://cloud.githubusercontent.com/assets/191598/15516063/0844eb18-21bf-11e6-97d3-196a60ec6bc0.png)

coreh commented:
We do. 🤔

Is it not showing for you?

<img width="342" alt="screen shot 2016-05-27 at 21 45 51" src="https://cloud.githubusercontent.com/assets/418473/15624052/91614f9e-2454-11e6-8421-b40aa434d408.png">

shaunandrews commented:
Huh, perhaps I was testing the wrong thing. The notice looks good when editing pages. Sorry for the confusion. Nice work 👍

coreh commented:
Now that #5681 is merged, we'll wait until the strings are translated and merge this.

Tug commented on client/post-editor/verify-email-dialog.jsx:
```diff
@@ -0,0 +1,118 @@
+import Dialog from 'components/dialog';
+import FormButton from 'components/forms/form-button';
+import Spinner from 'components/spinner';
+import i18n from 'lib/mixins/i18n';
```
Hi there, i18n usage has changed recently. please use:
```
import i18n from 'i18n-calypso';
```
instead.

coreh replied:
Fixed.

coreh commented:
Needs to be rebased to #5871.

coreh commented:
This is based on #5871. Once that is merged, I'll rebase this to master and then it will be ready to review.

apeatling commented:
One nitpick -- the small notice, could it have a more clear call to action? It is telling me to do something, without clearly telling me how -- unless I click learn more. What about if it said something more like "To publish, please check your email and verify your email address". Then I never have to click learn more.

folletto commented:
> What about if it said something more like "To publish, please check your email to verify your email address". Then I never have to click learn more.

Good point. I'd tweak it for something slightly shorter so at least in english will fit on two rows:

![screen shot 2016-06-20 at 12 40 57](https://cloud.githubusercontent.com/assets/4389/16192978/4e548462-36e4-11e6-89a7-8695a8d5c075.png)

coreh commented:
Rebased and squashed, the history was already really messy/nonsensical anyway.

Added the new wording, but changed the “to” to an “and”, I think it sounds better, since checking the email doesn't confirm the address, but is only one of the necessary steps. Thoughts?

![](https://cloudup.com/cwHTu5RWcMx+)

cc @meremagee

meremagee commented:
This is looking great as a first step in eliminating some awkward messaging and silent failures 🎉 

In the future, we may want to reduce the info text to something more concise, like:
> Please verify your email address.

> You should have received an email at xyz@xyz.com with a link to verify your email address when you registered for WordPress.com. If you can’t find the email, please click here to send it again.

> If the email address above is wrong, click here to change it in your account settings.

We may also want to change the verbiage from "click" to "tap" on touchscreens or think of something more generic for the inline link options.

Not blockers, though, as getting this out the door will be more helpful than bikeshedding!

michaeldcain commented on client/post-editor/editor-ground-control/index.jsx:
```diff
@@ -59,13 +64,34 @@ export default React.createClass( {
 			firstDayOfTheMonth: this.getFirstDayOfTheMonth(),
-			lastDayOfTheMonth: this.getLastDayOfTheMonth()
+			lastDayOfTheMonth: this.getLastDayOfTheMonth(),
+			needsVerification: config( 'email_verification_gate' ) && userUtils.needsVerificationForSite( this.props.site ),
```
As discussed in slack, using props in `getInitialState` can be an anti-pattern if the prop is updated after the component mounts. Until the user object is moved to redux, we discussed also updating the state on `componentWillReceiveProps`.

michaeldcain commented on client/post-editor/editor-ground-control/index.jsx:
```diff
@@ -91,10 +93,23 @@ export default React.createClass( {
 	},
 
+	componentWillReceiveProps: function( nextProps ) {
+		this.setState( {
```
Is there any benefit to only running `setState` if the `nextProps` differ from the current props?

`if ( nextProps.needsVerificationForSite( nextProps.site ) !== this.props.userUtils.needsVerificationForSite( this.props.site ) ) { }`

coreh replied:
I think setState already does the check, and only updates if it already has changed. Let me confirm that.

coreh replied:
https://facebook.github.io/react/docs/component-api.html

> `setState()` will always trigger a re-render unless conditional rendering logic is implemented in `shouldComponentUpdate()`. If mutable objects are being used and the logic cannot be implemented in `shouldComponentUpdate()`, calling `setState()` only when the new state differs from the previous state will avoid unnecessary re-renders.

So yeah I was wrong, it always triggers a re-render. Might make sense to add that check, then. (Thought I don't think this will be called frequently)

michaeldcain replied:
That's fair - we can just add it at a later point if needed.

aduth replied:
Why do we store this in state at all? Seems to fall within the same anti-pattern described in [_Props in getInitialState Is an Anti-Pattern_](https://facebook.github.io/react/tips/props-in-getInitialState-as-anti-pattern.html).

michaeldcain commented:
LGTM.

aduth commented on client/post-editor/editor-ground-control/index.jsx:
```diff
@@ -35,9 +37,12 @@ export default React.createClass( {
 		savedPost: React.PropTypes.object,
 		site: React.PropTypes.object,
+		user: React.PropTypes.object,
+		userUtils: React.PropTypes.object,
```
Why do you need to pass `userUtils` through as a prop instead of just calling as an import within this file?

aduth replied:
Further, can we do the same with `user`? As much as possible, avoid passing props from ancestors down to their children if the ancestors don't necessarily need to care about the concerns of that descendent component's behavior.

aduth commented on client/post-editor/editor-ground-control/index.jsx:
```diff
@@ -55,20 +60,71 @@ export default React.createClass( {
 	},
 
+	componentDidMount: function() {
+		if ( ! config( 'email_verification_gate' ) ) return;
```
>if/else/for/while/try blocks should always use braces, and always go on multiple lines.

https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#spacing

coreh replied:
👍 

That line is removed by #6230

aduth commented on client/post-editor/editor-ground-control/index.jsx:
```diff
@@ -55,20 +60,71 @@ export default React.createClass( {
+		if ( ! config( 'email_verification_gate' ) ) return;
+
+		this.props.user
+			&& this.props.user
```
```
./node_modules/.bin/eslint --quiet --rule 'operator-linebreak: 2' client/post-editor/editor-ground-control/index.jsx 

/Users/andrew/Documents/Code/wp-calypso/client/post-editor/editor-ground-control/index.jsx
   73:6  error  '&&' should be placed at the end of the line  operator-linebreak
   82:6  error  '&&' should be placed at the end of the line  operator-linebreak
   90:7  error  '&&' should be placed at the end of the line  operator-linebreak
   91:7  error  '&&' should be placed at the end of the line  operator-linebreak
  103:7  error  '&&' should be placed at the end of the line  operator-linebreak
  104:7  error  '&&' should be placed at the end of the line  operator-linebreak
  111:7  error  '&&' should be placed at the end of the line  operator-linebreak
  112:7  error  '&&' should be placed at the end of the line  operator-linebreak
  118:6  error  '&&' should be placed at the end of the line  operator-linebreak
  123:6  error  '&&' should be placed at the end of the line  operator-linebreak

✖ 10 problems (10 errors, 0 warnings)
```

coreh replied:
Yeah also fixing this on #6230

aduth commented on client/post-editor/editor-ground-control/index.jsx:
```diff
@@ -55,20 +60,71 @@ export default React.createClass( {
+
+	updateNeedsVerification: function() {
+		this.setState( {
+			needsVerification: config( 'email_verification_gate' )
```
When would we ever reach here that `config( 'email_verification_gate' )` is not true?

aduth commented on client/post-editor/editor-ground-control/index.jsx:
```diff
@@ -55,20 +60,71 @@ export default React.createClass( {
+	componentDidMount: function() {
+		if ( ! config( 'email_verification_gate' ) ) return;
+
+		this.props.user
```
You should include your `this.props.user` check in the early return line above, here and throughout the file.

aduth commented on client/post-editor/style.scss:
```diff
@@ -1,3 +1,49 @@
+.dialog.confirmation-dialog {
```
The class name for `confirmation-dialog` should be changed to respect CSS class naming guidelines, specifically:

> - The component fragment matches the folder name of the React component it's providing styles to.

aduth commented on client/post-editor/style.scss:
```diff
@@ -1,3 +1,49 @@
+	max-width: 450px;
+	border-radius: 3px;
+
+	h1 {
```
We suggest again tag name selectors in favor of classname using `.component__element` structure. This also helps avoid the need for nesting.

https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/css.md

aduth commented on client/post-editor/editor-ground-control/style.scss:
```diff
@@ -81,3 +81,54 @@
+	position: absolute;
+	left: 16px;
+	top: 16px;
+}
```
All files should end in a newline.

https://github.com/Automattic/wp-calypso/blob/0cfd9a6296bdbacf94bed4ec49d805c445c81c56/.editorconfig#L9

If you don't have one already, I suggest installing an [EditorConfig plugin](http://editorconfig.org/) for your editor.

aduth commented on client/post-editor/editor-ground-control/index.jsx:
```diff
@@ -399,6 +479,17 @@ export default React.createClass( {
+							className="editor-ground-control__email-verification-notice-icon" />
+						{ this.getVerificationNoticeLabel() }
+						{ ' ' }
+						<u>{ this.translate( 'Learn More' ) }</u>
```
Are you using `<u>` for an underline effect? That behavior was [deprecated in the HTML 4 specification](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/u). The correct semantic element here would be a `<button>`, which is [reset in our styles to remove default styling](https://github.com/Automattic/wp-calypso/blob/0cfd9a6296bdbacf94bed4ec49d805c445c81c56/client/components/button/style.scss#L6-L15).

coreh replied:
I was worried about the potential accessibility implications of using an  `<a>` or a  `<button>` element nested within an already clickable element (The entire notice is clickable, not only the “Learn More” portion).

I also felt like maybe adding a `<span>` with a class, and styling that would be overkill, that's why I went with `<u>`. I'm going to fix this on #6230.

Now that I think about it, we might need an `aria-role` for the notice, right?
